src/support.js

```diff
@@ -39,97 +36,64 @@ exports.theBest = (order, p1, p2) => {
     return p1[0]
   }
 
-  for (let firstCandidate of first) {
-    for (let secondCandidate of second) {
+  for (const firstCandidate of first) {
+    for (const secondCandidate of second) {
       if (firstCandidate === secondCandidate) {
         return firstCandidate
       }
     }
   }
 
-  throw new Error('No algorithms in common!')
+  throw new InvalidCryptoExchangeError('No algorithms in common!')
 }
 
-exports.makeMacAndCipher = (target, callback) => {
-  parallel([
-    (cb) => makeMac(target.hashT, target.keys.macKey, cb),
-    (cb) => makeCipher(target.cipherT, target.keys.iv, target.keys.cipherKey, cb)
-  ], (err, macAndCipher) => {
-    if (err) {
-      return callback(err)
-    }
+exports.makeMacAndCipher = async (target) => {
+  const macAndCipher = [ // they will run in parallel
+    makeMac(target.hashT, target.keys.macKey),
+    makeCipher(target.cipherT, target.keys.iv, target.keys.cipherKey)
+  ]
 
-    target.mac = macAndCipher[0]
-    target.cipher = macAndCipher[1]
-    callback()
-  })
+  target.mac = await macAndCipher[0] // ..but get awaited after each other
+  target.cipher = await macAndCipher[1]
```

Comment:
Any reason to not just do the following?

```js
[target.mac, target.cipher] = await Promise.all([
  makeMac(target.hashT, target.keys.macKey),
  makeCipher(target.cipherT, target.keys.iv, target.keys.cipherKey)
])
```

Reply:
Opinions. Which one is more readable in your opinion?

Reply:
The latter, imo, as I think it more clearly shows the parallelism, but this doesn't need to change. I think the second comment `// ..but get awaited after each other` might have been the thing that made me look at this twice.